### PR TITLE
elf2flt.c: add support for SOURCE_DATE_EPOCH

### PR DESCRIPTION
**Pull request by tpetazzoni**
The bFLT header has a "build_date" field which contains the date/time
at which the bFLT file was created. Unfortunately, this breaks
reproducible builds as two identical builds done at different times
will produce different results.

For example, on a bFLT binary, diffoscope reports the following
change:

  00000000: 6246 4c54 0000 0004 0000 0045 0001 48d4  bFLT.......E..H.
  00000010: 0002 85a0 0002 fe50 0000 1000 0002 85a0  .......P........
 -00000020: 0000 0757 0000 0001 5e05 742e 0000 0000  ...W....^.t.....
 +00000020: 0000 0757 0000 0001 5e05 a5c2 0000 0000  ...W....^.......

In order to address this, this commit adds support for the
SOURCE_DATE_EPOCH environment variable, which is standardized by the
reproducible-builds.org group at
https://reproducible-builds.org/specs/source-date-epoch/.

We simply use the time from the SOURCE_DATE_EPOCH variable (which
contains the number of seconds since Epoch) when SOURCE_DATE_EPOCH is
available in the environment, and otherwise fallback to the existing
logic that takes the current time using time().

Signed-off-by: Thomas Petazzoni <thomas.petazzoni@bootlin.com>

### CODE REVIEW

File: elf2flt.c [hunk 1]
```diff
@@ -1721,6 +1721,9 @@ int main(int argc, char *argv[])
 
   struct flat_hdr hdr;
 
+  char *sde;
```
Comment (vapier) [inline elf2flt.c]: `const char *`

File: elf2flt.c [hunk 2]
```diff
@@ -1948,7 +1951,14 @@ int main(int argc, char *argv[])
 	  | (pic_with_got ? FLAT_FLAG_GOTPIC : 0)
 	  | (docompress ? (docompress == 2 ? FLAT_FLAG_GZDATA : FLAT_FLAG_GZIP) : 0)
 	  );
-  hdr.build_date = htonl((uint32_t)time(NULL));
+
+  sde = getenv("SOURCE_DATE_EPOCH");
+  if (sde)
+      t = (time_t) strtoul(sde, NULL, 10);
```
Comment (vapier) [inline elf2flt.c]: no space between the cast & func call

is it even needed though ?  let the compiler figure it out ?

**Review by vapier — CHANGES_REQUESTED** [review]
binutils has a `--enable-deterministic-archives` configure flag.  imo we should add a similar `--enable-deterministic-flat` option (and perhaps even default it to true).

there is also a `D` flag to `ar` to control this at runtime.  we could add a `-D` flag to elf2flt to opt-in dynamically.

**Comment by tpetazzoni** [comment]
Thanks @vapier for you review. Do we really need a --enable-deterministic-archives compile-time option or a -D runtime option ? For example, gcc obeys to SOURCE_DATE_EPOCH, without any special option (it changes the behavior of the __DATE__ and __TIME__ macros).

**Comment by vapier** [comment]
i'm not saying we don't support SOURCE_DATE_EPOCH.  your change here (ignoring style nits) is fine.  i'm suggesting we add a configure option for it too.  it can be done in a follow up commit.

File: elf2flt.c (hunk 2 above)
Comment (yann-morin-1998) [inline elf2flt.c]: The toolchain wrapper in Buildroot use `strtoll()` and checks for `errno` after the attempted conversion.

Comment (yann-morin-1998) [reply elf2flt.c]: And as usual, reading the [docs](https://reproducible-builds.org/docs/source-date-epoch/) is insightful:
they use `strtoull()`.

**Comment by tpetazzoni** [comment]
I have pushed another iteration of the patch, which takes into account some of the comments. The (time_t) cast between strtoll() and the time_t variable is still needed, because strtoll will return a "signed long long", but the type of time_t may not necessarily match that.

**Comment by tpetazzoni** [comment]
Damn, just seen the comment from Yann suggesting to use strtoull(), so I'll do another iteration later.

**Comment by tpetazzoni** [comment]
Ok, here it comes, a new iteration. Hopefully this one will be good :-)

**Review by vapier — APPROVED** [review]

**Comment by yann-morin-1998** [comment]
As far as I'm concerned, this looks goo to me. Thanks. :-)

**Comment by vapier** [comment]
we really should have a configure flag as i suggested earlier though before [2038](https://en.wikipedia.org/wiki/Year_2038_problem) :p

**Comment by tpetazzoni** [comment]
Warning: I messed up and forgot to git amend before doing my Buildroot build, so I didn't actually test build the latest version. I'll do so later today, and come back with additional fixes if it turns out to be broken. Sorry about this :-/

Why do you think we should have a configure flag before 2038 ? The bFLT header only has 32-bit to store time already and my patch doesn't change anything to that, so the 2038 issue is not related to reproducible builds.

**Comment by vapier** [comment]
I'm not saying this change makes the situation worst, I'm saying the best way to avoid the overflow is to not use it at all. better to train people earlier rather than later that this field is not reliable so they don't get surprised when it does break.

**Pull request merged**